config/src/main/java/com/alibaba/nacos/config/server/service/HistoryService.java

```diff
@@ -114,4 +119,60 @@ private void checkHistoryInfoPermission(ConfigHistoryInfo configHistoryInfo, Str
             throw new AccessException("Please check dataId, group or namespaceId.");
         }
     }
+
+    /**
+     * Query the detailed configuration history information pair, including the original version and the updated version.
+     */
+    public ConfigHistoryInfoPair getConfigHistoryInfoPair(String dataId, String group, String namespaceId, Long nid)
+            throws AccessException {
+        ConfigHistoryInfo configHistoryInfo = historyConfigInfoPersistService.detailConfigHistory(nid);
+        if (Objects.isNull(configHistoryInfo)) {
+            return null;
+        }
+
+        // check if history config match the input
+        checkHistoryInfoPermission(configHistoryInfo, dataId, group, namespaceId);
+
+        // transform
+        ConfigHistoryInfoPair configHistoryInfoPair = new ConfigHistoryInfoPair();
+        BeanUtils.copyProperties(configHistoryInfo, configHistoryInfoPair);
+        configHistoryInfoPair.setOpType(configHistoryInfoPair.getOpType().trim());
+
+        if (OperationType.INSERT.getValue().equals(configHistoryInfoPair.getOpType())) {
+            configHistoryInfoPair.setUpdatedContent(configHistoryInfoPair.getContent());
+            configHistoryInfoPair.setUpdatedMd5(configHistoryInfoPair.getMd5());
+            configHistoryInfoPair.setContent(StringUtils.EMPTY);
+            configHistoryInfoPair.setMd5(StringUtils.EMPTY);
+        }
+
+        if (OperationType.UPDATE.getValue().equals(configHistoryInfoPair.getOpType())) {
+            ConfigHistoryInfo configHistoryInfoUpdated = historyConfigInfoPersistService.detailUpdatedConfigHistory(nid);
+            if (Objects.isNull(configHistoryInfoUpdated)) {
+                // get the latest config info
+                ConfigInfo configInfo = configInfoPersistService.findConfigInfo(dataId, group, namespaceId);
```

Comment:
should select the next version history, it may not be the lastest version

Reply:
> should select the next version history, it may not be the lastest version

The code in line 149 refers to querying the next historical version of the specified historical information indicated by nid.
The logic here is that when the operation type of the historical record is "U", but there is no next historical configuration record, it means that the changed version should be found in config_info instead of his_config_info.

Reply:
> should select the next version history, it may not be the lastest version

Supplement: For gray history config info of type 'U', if the next history config is empty,  query the latest version in TABLE‘config_info_gray’.